### PR TITLE
Uplift `clippy::invalid_utf8_in_unchecked` lint

### PR DESCRIPTION
**Pull request by Urgau**
This PR aims at uplifting the `clippy::invalid_utf8_in_unchecked` lint into two lints.

## `invalid_from_utf8_unchecked`

(deny-by-default)

The `invalid_from_utf8_unchecked` lint checks for calls to `std::str::from_utf8_unchecked` and `std::str::from_utf8_unchecked_mut` with an invalid UTF-8 literal.

### Example

```rust
unsafe {
    std::str::from_utf8_unchecked(b"cl\x82ippy");
}
```

### Explanation

Creating such a `str` would result in undefined behavior as per documentation for `std::str::from_utf8_unchecked` and `std::str::from_utf8_unchecked_mut`.

## `invalid_from_utf8`

(warn-by-default)

The `invalid_from_utf8` lint checks for calls to `std::str::from_utf8` and `std::str::from_utf8_mut` with an invalid UTF-8 literal.

### Example

```rust
std::str::from_utf8(b"ru\x82st");
```

### Explanation

Trying to create such a `str` would always return an error as per documentation for `std::str::from_utf8` and `std::str::from_utf8_mut`.

-----

Mostly followed the instructions for uplifting a clippy lint described here: https://github.com/rust-lang/rust/pull/99696#pullrequestreview-1134072751

@rustbot label: +I-lang-nominated
r? compiler

-----

For Clippy:

changelog: Moves: Uplifted `clippy::invalid_utf8_in_unchecked` into rustc

### CODE REVIEW

**Comment by lcnr** [comment]
r? compiler

**Comment by WaffleLapkin** [comment]
(I'll review code changes once T-lang approves the uplift)

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by lcnr** [comment]
@rust-lang/lang do you want this as two or as a single lint?

**Comment by nikomatsakis** [comment]
@rfcbot reviewef

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
My take is that two lints make sense; I agree with the premise that UB should be deny by default, but "mere" runtime errors not obviously so.

File: compiler/rustc_lint/src/invalid_from_utf8.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                    {
+                        lint(utf8_error);
+                    }
+                },
+                ExprKind::AddrOf(BorrowKind::Ref, _, Expr { kind: ExprKind::Array(args), .. }) => {
+                    let elements = args.iter().map(|e|{
+                        match &e.kind {
+                            ExprKind::Lit(Spanned { node: lit, .. }) => match lit {
+                                LitKind::Byte(b) => Some(*b),
+                                #[allow(clippy::cast_possible_truncation)]
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/invalid_from_utf8.rs:98]:
```suggestion
```

File: compiler/rustc_lint/messages.ftl [hunk 2]
```diff
@@ -536,3 +536,10 @@ lint_forget_ref = calls to `std::mem::forget` with a reference instead of an own
 lint_forget_copy = calls to `std::mem::forget` with a value that implements `Copy` does nothing
     .label = argument has type `{$arg_ty}`
     .note = use `let _ = ...` to ignore the expression or result
+
+# FIXME: we should ordinalize $valid_up_to when we support for that
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_lint/messages.ftl]: Oh, oops, for some reason I thought that was about pluralization, which *can* be handled: https://github.com/rust-lang/rust/blob/2fe47b966a8ee689d697583be4182262e7b4fd08/compiler/rustc_ast_passes/messages.ftl#L151-L157

Yeah, I'm also not sure what to do with ordinalization, so this fixme is fine.

File: compiler/rustc_lint/src/lints.rs [hunk 3]
```diff
@@ -699,6 +699,26 @@ pub struct ForgetCopyDiag<'a> {
     pub label: Span,
 }
 
+// invalid_from_utf8.rs
+
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/lints.rs:703]:
```suggestion
```

**Review by WaffleLapkin — APPROVED** [review]
The implementation looks good to me (modulo nit pick), now waiting for the FCP to properly complete (and a fix of git conflicts).

**Pull request merged**